### PR TITLE
Devpath for the back-end patterns

### PR DESCRIPTION
**Pull request by simon2k**
## Goal

The goal is of this dev path is to teach developers about patterns that can be used in the back-end web development. The current version is just an extraction of the existing patterns from RoR as a kick off for further adjustments.

### CODE REVIEW

**Comment by PoslinskiNet** [comment]
@kayinrage @mikoscz @stevo @qbart  feel free to provide feedback and your perspective [as TLs and/or Mastery seekers, Tech Evangelists, Tech Experts within Selleo around backend technologies/programming in general]

File: content/web_development/skills/Backend Patterns/advanced.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+**Points:** 1
+
+**Description:** You can apply the best practices across the framework while delivering solution.
+
+**Person who successfully completed requirement for given block can:**
+
+- Comprehend, design and deliver a full range of functionalities with no need for consultation (yet intuitively knows, when to ask for second opinion)
+- Demonstrate debugging skills for a full range of problems within application, also by investigating the framework code
+- Identify all framework capabilities and best-practices
+- Name, explain, choose and apply a wide range of software design patterns to solve particular problems
```
Comment (qbart) [inline content/web_development/skills/Backend Patterns/advanced.md:20]:
```suggestion
- Name, explain, choose and apply a wide range of software design patterns to solve problems
```

File: content/web_development/skills/Backend Patterns/basics.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+**Description:** You can apply the best practices across the framework while delivering solution.
+
+**Person who successfully completed requirement for given block can:**
+
+- Comprehend, design and deliver a full range of functionalities with no need for consultation (yet intuitively knows, when to ask for second opinion)
+- Demonstrate debugging skills for a full range of problems within application, also by investigating the framework code
+- Identify all framework capabilities and best-practices
+- Name, explain, choose and apply a wide range of software design patterns to solve particular problems
+- Demonstrate strong skills in TDD/BDD
```
Comment (qbart) [inline content/web_development/skills/Backend Patterns/basics.md:21]:
```suggestion
- Demonstrate good skills in TDD/BDD
```

File: content/web_development/skills/Backend Patterns/advanced.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+- "Microservices Oriented Architecture" - Compare it to monolith application. When to use? How communication between services looks like? What are the potential problems? What are the things that you should pay special attention to during design and implementation?
+- "JSON API vs GraphQL API vs REST API" - what are the pros and cons of using them? When will you suggest one over another?
+
+**OTHER QUESTIONS:**
+
+- Compare Client and Wrapper.
+- What is an Iterator? when should we use it, what problems does it solve for us?
+- What is a Value Object? when should we use it, what problems does it solve for us?
+- What is a Null Object, when should we use it, what problems does it solve for us?
+- What is Sanitization? When can it be useful?
```
Comment (qbart) [inline content/web_development/skills/Backend Patterns/advanced.md:56]: Sanitazation could be a part of basics

File: content/web_development/skills/Backend Patterns/basics.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+### 🎓 Learn
+
+- [Rails Guides](https://guides.rubyonrails.org/) (read parts that you have to use to create application described in katas below)
+- [RubyOnRails doctrine](https://rubyonrails.org/doctrine/)
+- [Thoughtbot - SOLID](https://thoughtbot.com/blog/back-to-basics-solid) AND you can watch movies as well:
+    - [Thoughtbot - SRP](https://thoughtbot.com/upcase/videos/single-responsibility-principle)
+    - [Thoughtbot - OCP](https://thoughtbot.com/upcase/videos/open-closed-principle)
+    - [Thoughtbot - LSP](https://thoughtbot.com/upcase/videos/liskov-substitution-principle)
+    - [Thoughtbot - ISP](https://thoughtbot.com/upcase/videos/interface-segregation-principle) (OPTIONAL)
+    - [Thoughtbot - DIP](https://thoughtbot.com/upcase/videos/dependency-inversion-principle) (OPTIONAL)
```
Comment (qbart) [inline content/web_development/skills/Backend Patterns/basics.md:38]: How about moving few of these to advanced?

**Review by qbart — CHANGES_REQUESTED** [review]
I've left some comments, for the first iteration it's pretty good - it would great to receive feedback from people.

File: content/web_development/skills/Backend Patterns/basics.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+**OTHER QUESTIONS:**
+
+- What is the naming convention in Rails?
+- Compare Active Record Pattern with Rails Active Record
+- What problems does Rails Migration help to mitigate?
+- What is the role of a Rails Model?
+- What is the role of a Rails Controller?
+- What is the role of a Rails View?
+- What is the role of a Rails Routes?
+- What are callbacks in Ruby on Rails?
```
[the remark this answers is not in the document]
Comment (simon2k) [reply content/web_development/skills/Backend Patterns/basics.md:71]: Thanks @qbart for the first review! I would like to make this framework/language independent (if it is possible). What do you think of it? FYI, this is a copy'n'paste extraction from our Rails dev path.

[the remark this answers is not in the document]
Comment (qbart) [reply content/web_development/skills/Backend Patterns/basics.md:71]: That was my initial assumption to make it technology agnostic but I've seen so many rails specific stuff that I thought it's for rails :smile: I'll go through this again and leave suggestions for making it that way

**Comment by PoslinskiNet** [comment]
@simon2k @qbart what is the status of this one?

**Comment by PoslinskiNet** [comment]
@simon2k @qbart any thoughts on this one?

**Comment by Skysplit** [comment]
bump

**Comment by PoslinskiNet** [comment]
@simon2k @qbart any thoughts on this one? This is my last call. I will close it automatically after a month from now if no response, since don't see any movement in this area.

**Comment by qbart** [comment]
I think we are lost here, I would close it for now and then create new one that is technology agnostic.

**Comment by simon2k** [comment]
It's gonna be re-created.

**Pull request closed without merging**